build_mt.py

```diff
@@ -1,599 +1,827 @@
+    merger_dir = merger_parent / sim_name
+    header = get_one_header(merger_dir)
+    
+    # simulation parameters
+    Lbox = header['BoxSize']
+    # location of the LC origins in Mpc/h
+    origins = np.array(header['LightConeOrigins']).reshape(-1,3)
+    origin = origins[0]
+
+    # directory where we save the final outputs
+    cat_lc_dir = Path("/mnt/gosling1/boryanah/light_cone_catalog/") / sim_name / "halos_light_cones/"
+    os.makedirs(cat_lc_dir, exist_ok=True)
+
+    # directory where we save the current state if we want to resume
+    os.makedirs(cat_lc_dir / "tmp", exist_ok=True)
+
+    # all redshifts, steps and comoving distances of light cones files; high z to low z
+    # LHG: these have to be recomputed per-sim. How is chi being determined? etaK from header?
+    zs_all = np.load("data_headers/redshifts.npy")
+    chis_all = np.load("data_headers/coord_dist.npy")
+    zs_all[-1] = float("%.1f" % zs_all[-1])  # LHG: I guess this is trying to match up to some filename or something?
+
+    # get functions relating chi and z
+    chi_of_z = interp1d(zs_all, chis_all)
+    z_of_chi = interp1d(chis_all, zs_all)
+
+    # more accurate, slightly slower
+    if not os.path.exists("data/zs_mt.npy"):
+        # all merger tree snapshots and corresponding redshifts
+        snaps_mt = sorted(merger_dir.glob("associations_z*.0.asdf"))
+        print(snaps_mt)
+        zs_mt = get_zs_from_headers(snaps_mt)
+        np.save("data/zs_mt.npy", zs_mt)
+    zs_mt = np.load("data/zs_mt.npy")
+
+
+    # fields are we extracting from the merger trees
+    # fields_mt = ['HaloIndex','HaloMass','Position','MainProgenitor','Progenitors','NumProgenitors']
+    # lighter version
+    fields_mt = ["HaloIndex", "Position", "MainProgenitor"]
+
+    # redshift of closest point on wall between original and copied box
+    z1 = z_of_chi(0.5 * Lbox - origin[0])
+    # redshift of closest point where all three boxes touch
+    # z2 = z_of_chi((0.5*Lbox-origin[0])*np.sqrt(2))
+    # furthest point where all three boxes touch; TODO: I think that's what we need
+    z2 = z_of_chi((0.5 * Lbox - origin[0]) * np.sqrt(3))
+
+    # corresponding indices
+    ind_start = np.argmin(np.abs(zs_mt - z_start))
+    ind_stop = np.argmin(np.abs(zs_mt - z_stop))
+
+    # initialize difference between the conformal time of last two shells
+    delta_chi_old = 0.0
+
+    # loop over each merger tree redshift
+    # LHG
+    # The algorithm: for each merger tree epoch, for each superslab, for each light cone origin,
+    # compute the intersection of the light cone with each halo, using the interpolated position
+    # to the previous merger epoch (and possibly a velocity correction).  If the intersection is
+    # between the current and previous merger epochs, then record the closer one as that halo's
+    # epoch and mark its progenitors as ineligible.
+    # Will need one padding superslab in the previous merger epoch.  Can process in a rolling fashion.
```

Comment:
I'm not totally clear on the algorithm below; top-level narrative would really help.  But I started to sketch one possible implementation, trying to minimize the amount of data that needs to be in memory at once and reduce copies of the box.